### DISCUSSION TITLE
Validate R peaks detection using knn outlier detection

### DISCUSSION THREAD

**Discussion by hernandezpachecodenis** [neuropsychology/NeuroKit#877] [Ideas]
Hello, it could be usefull to add as a misc function a procedure for r peaks detection validation. I've been using SPH ECG dataset and in some records the r peaks detected for each lead differs in the number, so when look in the whole r peaks dataframe the differences can be seen for the same interval of signal. I've trying a solution  but my python is no at the level requiered. 
I've been calculating, after detecting the outliers instance inside an r peaks dataframe, the euclidean distance of each element to the rest of normal instances and then finding the index of best ones, for droping the outliers elements in the outlier instance, until the length of it be equal to the normal legth of rest of normal instances of the same dataframe.
Also I've tried with another level of validation considering the distances beetwen RR intervals, in order to delete the r peaks whose distance in beetwen are the minor,  but although ECG is classified as normal this can differ in the number of samples in beetwen, and when look at the whole dataframe the values of coincident r peaks among ecg leads differs in its values. 
a more elegant solution could be using a knn outlier detector for each r peaks dataframe with outliers instance. Thanks

**Comment by danibene** [comment]
Hi @hernandezpachecodenis , thank you for sharing your thoughts here :) you said your python is not at the level required - how did you implement this evaluation on the SPH ECG dataset? I don't know if anyone currently has the time to implement and evaluate this, but I think it would be good if you could share the results of what you were able to evaluate (e.g. creating your own repository with the data, code, plots, etc.), so that if anyone wants to implement this eventually, they could build off of your work.

I imagine that this could be implemented as a function within `signal_fixpeaks()` https://github.com/neuropsychology/NeuroKit/blob/master/neurokit2/signal/signal_fixpeaks.py

Which could then also call a more general outlier-detection method in `find_outliers()`
https://github.com/neuropsychology/NeuroKit/blob/master/neurokit2/misc/find_outliers.py

I would recommend for you to explore the methods already available in `signal_fixpeaks()` and if possible try to compare them to the method you have in mind on open data!

**Reply by hernandezpachecodenis** [reply]
Hello, here is a sample of what I've done, ...I've been coding (I'm new in python) a more complex solution including some validation steps, in the ipynb are some explanations, I hope you can understand my poor english, if agree I can send the whole solution for one of the possibilities that can appear when detecting r peaks for differents leads and some of them differs in the quantity of the detected peaks.
Sorry the way I post results I have some limitations with the access to the internet
[Neurokit.zip](https://github.com/neuropsychology/NeuroKit/files/12337520/Neurokit.zip)
